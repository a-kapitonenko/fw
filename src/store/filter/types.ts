import { Frame } from '../frames/types';

export type Field = {
  name: string;
  value: string;
  checked?: boolean;
};

export type Groups = {
  color: Field[];
  width: Field[];
  noseBridge: Field[];
  shape: Field[];
  material: Field[];
};

export const enum FilterActionTypes {
  FETCH_REQUEST = '@@filter/FETCH_REQUEST',
  FETCH_SUCCESS = '@@filter/FETCH_SUCCESS',
  FETCH_ERROR = '@@filter/FETCH_ERROR',
  CHANGE_CHECKED = '@@filter/CHANGE_CHECKED',
};

export interface IFilterState {
  readonly fetching: boolean;
  readonly groups: Groups;
  readonly errors: string;
  readonly query: Object;
  readonly frames: Frame[];
};
