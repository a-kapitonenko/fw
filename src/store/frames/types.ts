export type Frame = {
  upc: number;
  label: string;
  value: string; 
  img: string;
  compatibility?: boolean;
};

export const enum FramesActionTypes {  
  SUBMIT_START = '@@frames/SUBMIT_START',
  SUBMIT_SUCCESS = '@@frames/SUBMIT_SUCCESS',
  SUBMIT_FAILED = '@@frames/SUBMIT_FAILED',
  FETCH_SIMILAR_FRAMES_START = '@@frames/FETCH_SIMILAR_FRAMES_START',
  FETCH_SIMILAR_FRAMES_SUCCESS = '@@frames/FETCH_SIMILAR_FRAMES_SUCCESS',
  FETCH_SIMILAR_FRAMES_FAILED = '@@frames/FETCH_SIMILAR_FRAMES_FAILED',
  CHECK_FRAME_START = '@@frames/CHECK_FRAME_START',
  CHECK_FRAME_SUCCESS = '@@frames/CHECK_FRAME_SUCCESS',
  CHECK_FRAME_FAILED = '@@frames/CHECK_FRAME_FAILED',
  OPEN = '@@frames/OPEN',
  CLOSE = '@@frames/CLOSE',
  SET_SELECTED_FRAME = '@@frames/SET_SELECTED_FRAME',
  CLEAR_SELECTED_FRAME = '@@frame/CLEAR_SELECTED_FRAME',
  CLEAR_RESULT = '@@frames/CLEAR_RESULT',
};

export interface FramesState {
  readonly isFetching: boolean;
  readonly errors: string;
  readonly open: boolean;
  readonly selectedFrame: Frame;
  readonly similarFrames: {
    readonly isFetching: boolean;
    readonly errors: string;
    readonly data: Frame[];
  };
};
