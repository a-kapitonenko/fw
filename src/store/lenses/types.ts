export type Lens = {
  name: string;
  value: number;
};

export const enum LensesActionTypes {
  FETCH_LENSES_START = '@@lenses/FETCH_LENSES_START',
  FETCH_LENSES_SUCCESS = '@@lenses/FETCH_LENSES_SUCCESS',
  FETCH_LENSES_FAILED = '@@lenses/FETCH_LENSES_FAILED',

  SAVE_LENS_START = '@@lenses/SAVE_LENS_START',
  SAVE_LENS_SUCCESS = '@@lenses/SAVE_LENS_SUCCESS',
  SAVE_LENS_FAILED = '@@lenses/SAVE_LENS_FAILED',

  CHECK_LENS_START = '@@lenses/CHECK_LENS_START',
  CHECK_LENS_SUCCESS = '@@lenses/CHECK_LENS_SUCCESS',
  CHECK_LENS_FALIED = '@@lenses/CHECK_LENS_FAILED',
};

export interface ILensesState {
  readonly isFetching: boolean;
  readonly errors: string;
  readonly lenses: Lens[];
};
