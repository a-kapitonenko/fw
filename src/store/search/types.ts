import { Frame } from '../frames/types';

export const enum SearchActionTypes {
  FETCH_REQUEST = '@@search/FETCH_REQUEST',
  CLOSE_REQUEST = '@@search/CLOSE_REQUEST',
  SET_ERRORS = '@@search/SET_ERRORS',
  CLEAR_ERRORS = '@@search/CLEAR_ERRORS',
  FETCH_FRAMES = '@@search/FETCH_FRAMES',
  SET_FRAMES = '@@search/SET_FRAMES',
  SET_SELECTED_FRAMES = '@@search/SET_SELECTED_FRAMES',
}

export interface ISearchState {
  readonly isFetching: boolean;
  readonly errors: string;
  readonly frames: Frame[];
  readonly selectedFrames: Frame[];
}
