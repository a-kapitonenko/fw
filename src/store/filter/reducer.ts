import { Reducer } from 'redux';
import { IFilterState, FilterActionTypes, Groups, Field } from './types';

const initialGroupsState = {
  color: [],
  width: [],
  noseBridge: [],
  shape: [],
  material: []
}

const initialState: IFilterState = {
  isFetching: false,
  errors: '',
  query: initialGroupsState,
  frames: [],
  groups: {
    isFetching: false,
    errors: '',
    data: <Groups>initialGroupsState,
  }
};

const reducer: Reducer<IFilterState> = (state = initialState, action) => {
  switch (action.type) {
    case FilterActionTypes.FILTERING_START: {
      return { ...state, isFetching: true, errors: '' };
    }
    case FilterActionTypes.FILTERING_SUCCESS: {
      return { ...state, isFetching: false, frames: action.payload };
    }
    case FilterActionTypes.FILTERING_FAILED: {
      return { ...state, isFetching: false, errors: action.payload };
    }
    case FilterActionTypes.FETCH_GROUPS_START: {
      return { ...state, groups: { ...state.groups, isFetching: true, errors: '' } };
    }
    case FilterActionTypes.FETCH_GROUPS_SUCCESS: {
      return { ...state, groups: { ...state.groups, isFetching: false, data: action.payload } };
    }
    case FilterActionTypes.FETCH_GROUPS_FAILED: {
      return { ...state, groups: { ...state.groups, isFetching: true, errors: action.paylaod } };
    }
    case FilterActionTypes.CHANGE_CHECKED: {
      const item = state.groups.data[action.payload.type].map((field: Field) => {
        if (field.name === action.payload.name) {
          return { ...field, checked: action.payload.value };
        }

        return { ...field };
      });

      return {
        ...state,
        groups: {
          ...state.groups,
          data: {
            ...state.groups.data,
            [action.payload.type]: item
          }
        }
      };
    }
    case FilterActionTypes.CLEAR_CHECKED: {
      const groups = <Groups>{};

      for (const group in state.groups.data) {
        groups[group] = state.groups.data[group].map((field: Field) => ({ ...field, checked: false }));
      }

      return { ...state, groups: { ...state.groups, data: groups } };
    }
    case FilterActionTypes.ADD_QUERY: {
      return {
        ...state,
        query: {
          ...state.query,
          [action.payload.type]: [
            ...state.query[action.payload.type],
            action.payload.value
          ]
        }
      };
    }
    case FilterActionTypes.DELETE_QUERY: {
      const index = state.query[action.payload.type].indexOf(action.payload.value);

      return {
        ...state,
        query: {
          ...state.query,
          [action.payload.type]: [
            ...state.query[action.payload.type].slice(0, index),
            ...state.query[action.payload.type].slice(index + 1)
          ]
        }
      };
    }
    case FilterActionTypes.CLEAR_QUERY: {
      return { ...state, query: initialGroupsState };
    }
    default: {
      return state;
    }
  }
};

export { reducer as FilterReducer };
