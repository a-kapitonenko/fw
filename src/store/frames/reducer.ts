import { Reducer } from 'redux';
import { FramesState, FramesActionTypes, Frame } from './types';

const initialState: FramesState = {
  isFetching: false,
  errors: '',
  open: false,
  selectedFrame: <Frame>{},
  similarFrames: {
    isFetching: false,
    errors: '',
    data: []
  },
};

const reducer: Reducer<FramesState> = (state = initialState, action) => {
  switch (action.type) {
    case FramesActionTypes.FETCH_SIMILAR_FRAMES_START: {
      return { ...state, similarFrames: { ...state.similarFrames, isFetching: true, errors: '' } };
    }
    case FramesActionTypes.FETCH_SIMILAR_FRAMES_SUCCESS: {
      return { ...state, similarFrames: { ...state.similarFrames, isFetching: false, data: action.payload } };
    }
    case FramesActionTypes.FETCH_SIMILAR_FRAMES_FAILED: {
      return { ...state, similarFrames: { ...state.similarFrames, isFetching: false, errors: action.payload } };
    }
    case FramesActionTypes.SUBMIT_START: {
      return { ...state, isFetching: true, errors: '' };
    }
    case FramesActionTypes.SUBMIT_SUCCESS: {
      return { ...state, isFetching: false };
    }
    case FramesActionTypes.SUBMIT_FAILED: {
      return { ...state, isFetching: false, errors: action.payload };
    }
    case FramesActionTypes.OPEN: {
      return { ...state, open: true };
    }
    case FramesActionTypes.CLOSE: {
      return { ...state, open: false };
    }
    case FramesActionTypes.SET_SELECTED_FRAME: {
      return { ...state, selectedFrame: action.payload };
    }
    case FramesActionTypes.CLEAR_SELECTED_FRAME: {
      return { ...state, selectedFrame: <Frame>{} };
    }
    default: {
      return state;
    }
  }
};

export { reducer as FramesReducer };
