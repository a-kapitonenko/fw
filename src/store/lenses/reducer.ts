import { Reducer } from 'redux';
import { LensesState, LensesActionTypes } from './types';

const initialState: LensesState = {
  fetching: false,
  list: [],
  errors: '',
}

const reducer: Reducer<LensesState> = (state = initialState, action) => {
  switch (action.type) {
    case LensesActionTypes.FETCH_REQUEST: {
      return { ...state, fetching: true };
    }
    case LensesActionTypes.FETCH_SUCCESS: {
      return { 
        ...state, 
        fetching: false, 
        list: action.payload 
      };
    }
    case LensesActionTypes.FETCH_ERROR: {
      return { 
        ...state, 
        fetching: false, 
        errors: action.payload 
      };
    }
    default: {
      return state
    }
  }
};

export { reducer as LensesReducer }
