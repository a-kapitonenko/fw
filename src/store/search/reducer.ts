import { Reducer } from 'redux';
import { ISearchState, SearchActionTypes } from './types';

const initialState: ISearchState = {
  isFetching: false,
  errors: '',
  frames: [],
  selectedFrames: []
};

const reducer: Reducer<ISearchState> = (state = initialState, action) => {
  switch (action.type) {
    case SearchActionTypes.SEARCH_START: {
      return { ...state, isFetching: true, errors: '' };
    }
    case SearchActionTypes.SEARCH_SUCCESS: {
      return { ...state, isFetching: false, frames: action.payload };
    }
    case SearchActionTypes.SEARCH_FAILED: {
      return { ...state, isFetching: false, errors: action.paylaod };
    }
    case SearchActionTypes.SET_SELECTED_FRAMES: {
      return { ...state, selectedFrames: action.payload };
    }
    default: {
      return state;
    }
  }
};

export { reducer as SearchReducer };
