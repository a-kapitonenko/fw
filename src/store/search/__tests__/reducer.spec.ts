import { SearchActionTypes } from '../types';
import { Frame } from '../../frames/types';
import { SearchReducer } from '../reducer';
import * as mockData from '../../../mockData';

describe('search reducer', () => {
  it('test handles SEARCH_START', () => {
    const searchAction = {
      type: SearchActionTypes.SEARCH_START,
      payload: {},
    };
    
    expect(SearchReducer(undefined, searchAction)).toMatchSnapshot();
  });

  it('test handles SEARCH_SUCCESS', () => {
    const searchPayload: Frame[] = mockData.frames;

    const searchAction = {
      type: SearchActionTypes.SEARCH_SUCCESS,
      payload: searchPayload,
    };

    expect(SearchReducer(undefined, searchAction)).toMatchSnapshot();
  });

  it('test handles SEARCH_FAILED', () => {
    const searchPayload: string = mockData.errorMessage;

    const searchAction = {
      type: SearchActionTypes.SEARCH_FAILED,
      payload: searchPayload,
    };

    expect(SearchReducer(undefined, searchAction)).toMatchSnapshot();
  });

  it('test handles SET_SELECTED_FRAMES', () => {
    const searchPayload: Frame[] = mockData.frames;

    const searchAction = {
      type: SearchActionTypes.SET_SELECTED_FRAMES,
      payload: searchPayload,
    };

    expect(SearchReducer(undefined, searchAction)).toMatchSnapshot();
  });

  it('test handles CLEAR_RESULT', () => {
    const searchAction = {
      type: SearchActionTypes.CLEAR_RESULT,
      payload: {},
    };

    expect(SearchReducer(undefined, searchAction)).toMatchSnapshot();
  });
});
