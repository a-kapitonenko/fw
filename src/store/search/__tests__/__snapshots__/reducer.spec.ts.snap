// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search reducer test handles CLEAR_RESULT 1`] = `
Object {
  "errors": "",
  "frames": Array [],
  "isFetching": false,
  "selectedFrames": Array [],
}
`;

exports[`search reducer test handles SEARCH_FAILED 1`] = `
Object {
  "errors": "Error message",
  "frames": Array [],
  "isFetching": false,
  "selectedFrames": Array [],
}
`;

exports[`search reducer test handles SEARCH_START 1`] = `
Object {
  "errors": "",
  "frames": Array [],
  "isFetching": true,
  "selectedFrames": Array [],
}
`;

exports[`search reducer test handles SEARCH_SUCCESS 1`] = `
Object {
  "errors": "",
  "frames": Array [
    Object {
      "compatibility": true,
      "img": "1.png",
      "label": "FM17116",
      "upc": 82850553738,
      "value": "82850553738",
    },
    Object {
      "compatibility": false,
      "img": "2.png",
      "label": "SA17346",
      "upc": 82850523999,
      "value": "82850523999",
    },
  ],
  "isFetching": false,
  "selectedFrames": Array [],
}
`;

exports[`search reducer test handles SET_SELECTED_FRAMES 1`] = `
Object {
  "errors": "",
  "frames": Array [],
  "isFetching": false,
  "selectedFrames": Array [
    Object {
      "compatibility": true,
      "img": "1.png",
      "label": "FM17116",
      "upc": 82850553738,
      "value": "82850553738",
    },
    Object {
      "compatibility": false,
      "img": "2.png",
      "label": "SA17346",
      "upc": 82850523999,
      "value": "82850523999",
    },
  ],
}
`;
